R/appMetadata.R

```diff
@@ -0,0 +1,378 @@
+  # to be run by rmarkdown::run (rmd-shiny).
+  serverR <- grep("^server.r$", appFiles, ignore.case = TRUE, perl = TRUE)
+  if (length(serverR) > 0) {
+    return("shiny")
+  }
+
+  # Any non-Shiny R Markdown or Quarto documents are rendered content and get
+  # rmd-static or quarto-static.
+  if (length(rmdFiles) > 0 || length(qmdFiles) > 0) {
+    if (!is.null(quartoInfo)) {
+      return("quarto-static")
+    } else {
+      # For Shinyapps and posit.cloud, treat "rmd-static" app mode as "rmd-shiny" so that
+      # they can be served from a shiny process in Connect until we have better support of
+      # rmarkdown static content
+      if (isCloudServer) {
+        return("rmd-shiny")
+      }
+      return("rmd-static")
+    }
+  }
+
+  # We don't have an RMarkdown, Shiny app, or Plumber API, but we have a saved model
+  if (length(grep("(saved_model.pb|saved_model.pbtxt)$", appFiles, ignore.case = TRUE, perl = TRUE)) > 0) {
+    return("tensorflow-saved-model")
+  }
+
+  # no renderable content here; if there's at least one file, we can just serve
+  # it as static content
+  if (length(appFiles) > 0) {
+    return("static")
+  }
+
+  cli::cli_abort(
+    "No content to deploy; {.arg appFiles} is empty"
+  )
+}
+
+isShinyRmd <- function(filename) {
+  yaml <- yamlFromRmd(filename)
+  if (is.null(yaml)) {
+    return(FALSE)
+  }
+  is_shiny_prerendered(yaml$runtime, yaml$server)
+}
+# Adapted from rmarkdown:::is_shiny_prerendered()
+is_shiny_prerendered <- function(runtime, server = NULL) {
+  if (!is.null(runtime) && grepl("^shiny", runtime)) {
+    TRUE
+  } else if (identical(server, "shiny")) {
+    TRUE
+  } else if (is.list(server) && identical(server[["type"]], "shiny")) {
+    TRUE
+  } else {
+    FALSE
+  }
+}
+
+# If deploying an R Markdown, Quarto, or static content, infer a primary
+# document if one is not already specified.
+  # Note: functionality in inferQuartoInfo() depends on primary doc inference
+  # working the same across app modes.
+inferAppPrimaryDoc <- function(appPrimaryDoc, appFiles, appMode) {
+  if (!is.null(appPrimaryDoc)) {
+    return(appPrimaryDoc)
+  }
+
+  # Only documents and static apps don't have primary _doc_
+  if (!(appIsDocument(appMode) || appMode == "static")) {
+    return(appPrimaryDoc)
+  }
+
+  # determine expected primary document extension
+  ext <- if (appMode == "static") "\\.html?$" else "\\.[Rq]md$"
+
+  # use index file if it exists
+  matching <- grepl(paste0("^index", ext), appFiles, ignore.case = TRUE)
+  if (!any(matching)) {
+    # no index file found, so pick the first one we find
+    matching <- grepl(ext, appFiles, ignore.case = TRUE)
+
+    if (!any(matching)) {
+      cli::cli_abort(c(
+        "Failed to determine {.arg appPrimaryDoc}.",
+        x = "No files matching {.str {ext}}."
+      ))
+    }
+  }
+
+  if (sum(matching) > 1) {
+    # if we have multiple matches, pick the first
+    appFiles[matching][[1]]
```

Comment:
This "alphabetically first" heuristic probably deserves to be explained, somewhere.

Reply:
I wonder if it shouldn't actually be an error

Reply:
This flow happens when trying to deploy an entire directory containing `a.Rmd` and `b.Rmd`. This choice is "reasonable", but it was not an explicit user choice.

This is effectively the behavior of:

```r
rmarkdown::run(file = NULL, dir = getwd())
```

That said, it would be better to encourage folks to make an explicit choice.

Reply:
The other problem is I have no mental model of what "primary" doc actually means.

Reply:
The simplest example: Assume you have a directory layout like:

```
project/
    primary.Rmd
    chunka.Rmd
    chunkb.Rmd
```

The `rmarkdown::render` needs to know its rendering target. It cannot be called against a directory. The primary document in this case is `primary.Rmd`.

Personally, I always give my projects meaningful names and then have `index.Rmd` within it, which makes the notion of "primary" pretty obvious. More complicated layouts are more common, however, and we often see folks with collections of Rmd files, and they all get deployed together, even though they often want only one document or a subset of the Rmds included.

Many other types of content do not require a primary document; this rendering case is the strongest example.